detectron2/engine/hooks.py

```diff
@@ -199,6 +201,81 @@ def after_step(self):
         self.step(self.trainer.iter)
 
 
+class BestCheckpointer(HookBase):
+    """
+    Checkpoints best weights based off given metric.
+
+    This hook should be used in conjunction to and executed after `EvalHook`.
+    """
+
+    def __init__(
+        self,
+        eval_period: int,
+        checkpointer: Checkpointer,
+        val_metric: str = "bbox/AP50",
+        file_prefix: str = "model_best",
+        append_iter: bool = False,
+    ) -> None:
+        """
+        Args:
+            eval_period (int): the period `EvalHook` is set to run.
+            checkpointer: the checkpointer object used to save checkpoints.
+            val_metric (str): chosen validation metric to track for best checkpointing
+            file_prefix (str): the prefix of checkpoint's filename, defaults to "model_best"
+            append_iter (bool): if True, checkpoint file name will have iteration num appended, else write to the same checkpoint file.
+        """
+        self._logger = logging.getLogger(__name__)
+        self._period = eval_period
+        self._val_metric = val_metric
+        self._checkpointer = checkpointer
+        self._file_prefix = file_prefix
+        self.best_metric = None
+        self.best_iter = None
+        self._append_iter = append_iter
+
+    def _best_checking(self):
+        latest_metric, metric_iter = self.trainer.storage.latest().get(self._val_metric)
+        if self.best_metric is None or latest_metric > self.best_metric:
+            if self._append_iter:
+                additional_state = {"iteration": metric_iter}
+                save_name = f"{self._file_prefix}_{metric_iter:07d}"
+            else:
+                save_name = f"{self._file_prefix}"
+                additional_state = {}
+            self._checkpointer.save(save_name, **additional_state)
+            if self.best_metric is None:
+                self._logger.info(
+                    f"Saved first model with latest eval score for {self._val_metric} at {latest_metric:0.5f}"
+                )
```

Comment:
Saving the first model looks useless. Could you change the behavior to not save?

Reply:
I've modified it such that it only updates the `best_metric` during the "first model" but does not save it. 

However, I do have some lingering concerns for this -- should we just save it as long as the current metric value is valid (not NaN nor infs), even if it is the first model? Seems like it is implemented this way in Keras as well, as seen [here](https://github.com/keras-team/keras/blob/f43c792c129ffba27943c00f11c73c7d8db2838f/keras/callbacks.py#L1412). It initialises the best variable as either negative or positive infinity, then compares with current value and saves even if it is the "first call". We could implement that in a similar fashion here.

Reply:
Would it be a good idea to add a starting threshold as a keyword argument defaulted to None?
Often when I train models on my dataset I know what to expect from the metric and if it performs a lot worse then I don't want to save it.

Reply:
@ppwwyyxx I do realise that sometimes saving the model after the first evaluation is needed. It is always possible the best model is reached by the first evaluation run, albeit not ideal, and especially if the first evaluation happens after many steps. 

I will pull this commit into this PR if you feel it is alright: https://github.com/levan92/detectron2/commit/59bebb201f59519e8db80f81c0ad2f06347053e0 It exposes a `save_first` argument which default to `False`.

Reply:
That argument makes total sense to me. We can let it be the default behavior (i.e. also save the first evaluation).